examples/blocked_cholesky_multigpu.py

```diff
@@ -197,68 +205,76 @@ async def test_blocked_cholesky():
-    for d in range(num_gpus):
-      row_size = n // (block_size * num_gpus)
-      if d < ((n / block_size) % num_gpus):
-        row_size += 1
-      if row_size > 0:
-        allocate_gpu_memory(d, row_size, n, block_size)
-    logger.debug("Allocate memory done..")
-
-    np.random.seed(10)
-
-    logger.debug("Random number generate..")
-    # Construct input data
-    a = np.random.rand(n, n)
-    a = a @ a.T
-    logger.debug("Random number generate done..")
-
-    logger.debug("Copy a to a1..")
-    # Copy and layout input
-    a1 = a.copy()
-    logger.debug("Copying done..")
-    logger.debug("Shaping starts..")
-    ap = a1.reshape(n // block_size, block_size, n // block_size, block_size).swapaxes(1,2)
-    logger.debug("Shaping done..")
-    set_gpu_memory_from_cpu(ap, num_gpus)
-
-    for i in range(len(gpu_arrs)):
-      logger.debug("Device ", i, " arrays are on ", gpu_arrs[i].device);
-    logger.debug("Calculate starts..")
-    start = time.perf_counter()
-
-    # Call Parla Cholesky result and wait for completion
-    await cholesky_blocked_inplace(ap, num_gpus)
-
-    end = time.perf_counter()
-    print(end - start, "seconds")
-    logger.debug("Calculate done..")
-
-    for i in range(len(gpu_arrs)):
-      logger.debug("Device ", i, " arrays are on ", gpu_arrs[i].device);
-
-    for d in range(len(gpu_arrs)):
-      logger.debug("Device:", d, " swap array..")
-      gpu_arrs[d] = cp.swapaxes(gpu_arrs[d], 2, 1)
-      logger.debug("Device:", d, " swap done..")
-
-    cpu_arrs = np.empty([n // block_size,
-                         block_size,
-                         n // block_size,
-                         block_size], dtype=float)
-    for r_num in range(n // block_size):
-      dev_id   = r_num % num_gpus
-      local_id = r_num // num_gpus
-      cpu_arrs[r_num] = cp.asnumpy(gpu_arrs[dev_id][local_id])
-    cpu_arrs = cpu_arrs.reshape(n, n)
-    print("Truth", linalg.cholesky(a).T)
-
-    # Check result
-    computed_L = np.tril(cpu_arrs)
-    print("Soln", computed_L)
-    error = np.max(np.absolute(a-computed_L @ computed_L.T))
-    print("Error", error)
-    assert(error < 1E-8)
+    for i in range(num_tests):
+      global gpu_arrs
+      gpu_arrs = []
+      for d in range(num_gpus):
+        row_size = n // (block_size * num_gpus)
+        if d < ((n / block_size) % num_gpus):
+          row_size += 1
+        if row_size > 0:
+          allocate_gpu_memory(d, row_size, n, block_size)
+      logger.debug("Allocate memory done..")
+
+      np.random.seed(10)
+
+      logger.debug("Random number generate..")
+      # Construct input data
+      a = np.random.rand(n, n)
+      a = a @ a.T
+      logger.debug("Random number generate done..")
+
+      logger.debug("Copy a to a1..")
+      # Copy and layout input
+      a1 = a.copy()
+      logger.debug("Copying done..")
+      logger.debug("Shaping starts..")
+      ap = a1.reshape(n // block_size, block_size, n // block_size, block_size).swapaxes(1,2)
```

Comment:
This is just a comment, otherwise it lgtm

iirc  this takes a lot of time to setup the matrix but I'm not sure what % of total runtime it is. 

A quality of life improvement for faster timing runs might be setting it up once at the beginning and creating copies of the data instead. (To avoid losing the GPU memory due to the extra copy you could store them in host and copy back at the start of a loop)

Reply:
Good point. I will fix this :-)